File: Golang/628. Maximum Product of Three Numbers.go
func maximumProduct(nums []int) int {
    num3 := []int(nums[:3])
    sort.Ints(num3)
    min1, min2 := num3[1], num3[1]
    max1, max2, max3 := num3[0], num3[0], num3[0]
    
    for _, num := range nums {
        if num > max1 {
            max1, max2, max3 = num, max1, max2
        } else if num > max2 {
            max2, max3 = num, max2
        } else if num > max3 {
            max3 = num
        }
        
        if num < min1 {
            min1, min2 = num, min1
        } else if num < min2 {
            min2 = num
        }
    }
    
    return max(min1*min2*max1, max1*max2*max3)
}

func max(a, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

File: Python3/628. Maximum Product of Three Numbers.py
class Solution:
    def maximumProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # nums = sorted(nums) # 108ms
        # min1, min2, max1, max2, max3 = nums[0], nums[1], nums[-1], nums[-2], nums[-3]
        # return max(min1*min2*max1, max1*max2*max3)
        
        num3 = sorted(nums[:3])
        min1 = min2 = num3[1]
        max3 = max2 = max1 = num3[0]
        for num in nums:
            if num > max1:
                max1, max2, max3 = num, max1, max2
            elif num > max2:
                max2, max3 = num, max2
            elif num > max3:
                max3 = num
            else:
                pass

            if num < min1:
                min1, min2 = num, min1
            elif num < min2:
                min2 = num
            else:
                pass
        return max(min1*min2*max1, max1*max2*max3)
